README.md
# Assignment-Fetcher
This is a vscode extension that provides functionalities on assignments following specific structure and fetched using github classroom. Functionalities include - starting and switching off containers associated with the project directory, Track and store active code time, auto run code in container & commit on save, forced auto save according to specified interval, push to repository using a button. 

## Requirements
1. Git and github classroom ewxtension should be installed and working with vscode.
2. Docker & Docker-Compose should be installed and working without sudo command.
3. Extension settings should be filled completely in vscode.
4. The initial directory for assignment repository should be fetched and amde using github classroom.
5. The assignment folder/repostory should follow the strict structure of having the `Data` folder immediately inside the assigment folder/repository with `docker-compose.yml`, `Dockerfile`, `studentData.json` with `"activeTime"` key set to `0` beforehand and `metadata.json` with `"container_name"` set beforehand.

## Explicit Features
1. `Start container`: starts a container using docker-compose.yml.
2. `Submit Progress`: pushes to the repository set up by github classroom.
3. `Run`: runs the code in the container.

`2` and `3` will be visible only after starting the container and will be hidden otherwise.

## Implicit Features
1. `Auto run and commit`: each save will trigger run and commit silently.
2. `Active code time`: amount of time spent coding which stops after each inactivity of duration specified by `Timer Shutdown Interval`.
3. `Forced Save after predefined intervals`: after each duration specified by `Force Save Interval` all files inside the assignment folder will be saved.

## Extension Settings
1. `Force Save Interval`: Minutes after which files will be automatically saved periodically. 
2. `Timer Shutdown Interval`: Minutes after which user is marked inactive.

After changing settings, restarting vscode is encouraged.

CHANGELOG.md
# Change Log

All notable changes to the "assignmentfetcher" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

media/main.js
//  @ts-noCheck

// This script will be run within the webview itself
// It cannot access the main VS Code APIs directly.
// console logs arent emitted here !?
(function () {
    // @ts-ignore
    const vscode = acquireVsCodeApi();
    // const oldState = vscode.getState() || { switchState: 'off' };
    var powerBtn = document.getElementsByClassName('switch')[0];
    var check = document.getElementById('check');
    var indicator = document.getElementById('indicator');
    var submitBtn = document.getElementById('submitProgress');
    var runBtn = document.getElementById('run');
    var actions = document.getElementById('actions');

    vscode.postMessage({ type: 'getState'});
    window.addEventListener('message', event => {
        const message = event.data; // The json data that the extension sent
        switch (message.type) {
            case 'getState':{
                if(message.isInitialized){
                    check.checked = true;
                    submitBtn.disabled = false;
                    runBtn.disabled = false;

                    actions.style.visibility = "visible";
                    actions.style.width = "100%";
                    actions.style.opacity = "100%";
                    indicator.textContent = 'ON';
                    submitBtn.style.cursor = "pointer"; 
                    runBtn.style.cursor = "pointer"; 
                }
                else{
                    check.checked = false;
                    submitBtn.disabled = true;
                    runBtn.disabled = true;

                    actions.style.width = "0%"
                    actions.style.opacity = "0%"
                    indicator.textContent = 'OFF';
                    submitBtn.style.cursor = "default"; 
                    runBtn.style.cursor = "default"; 
                }
                break;
            }
        }
    });

    powerBtn.addEventListener('click', function(e) {
        e.preventDefault();
        console.log('call to backend initialize')
        vscode.postMessage({type: 'initialize'});
    });
    
    submitBtn.addEventListener('click', () => {
        vscode.postMessage({ type: 'submitProgress'});
    });

    runBtn.addEventListener('click', () => {
        vscode.postMessage({ type: 'run'});
    });

}());

src/actions.ts
import * as vscode from 'vscode';
// import { storeTime } from './storeHandler';
// import { GitHandler } from './gitHandler';
// import { DockerHandler } from './dockerHandler';

// const path = require("path");
// const nconf = require('nconf')
// const os = require('os');

// let Git: GitHandler;
// let Docker: DockerHandler;
// let folderPath: string;
// let startTime:number;
// let latestTime:number;
// let timerID:null|NodeJS.Timeout = null;
// let saveProgresstimerID:null|NodeJS.Timeout = null;
// let forcedSaveTimerID:null|NodeJS.Timeout = null;

// let isInitialized: boolean = false;
// let fileChangeListenerDisposable: vscode.Disposable|null = null;
// let fileSaveListenerDisposable: vscode.Disposable|null = null;
// // export { folderPath }

// export function initialize(){
// 	if(isInitialized){
// 		vscode.window.showInformationMessage('Already Initialized');
// 		return;
// 	}
// 	if(vscode.workspace.workspaceFolders !== undefined) {

// 		folderPath = vscode.workspace.workspaceFolders[0].uri.fsPath;
// 		Git = new GitHandler(folderPath);
// 		nconf.use('file', { file: path.join(folderPath, 'Data', 'metadata.json') });
// 		nconf.load();
// 		Docker = new DockerHandler(folderPath, nconf.get("container_name"));

// 		vscode.window.withProgress({
// 			cancellable: false,
// 			location: vscode.ProgressLocation.Notification,
// 			title: "Starting Container"
// 		},	async (progress) => {
// 			try{
// 				let response = await Docker.exec('start');
// 				const fswatcher = vscode.workspace.createFileSystemWatcher(new vscode.RelativePattern(folderPath, '*.cpp')); 	
// 				fileChangeListenerDisposable = fswatcher.onDidChange(onFileChangeHandler);
// 				fileSaveListenerDisposable = vscode.workspace.onDidSaveTextDocument(onSaveHandler);
// 				intervalForceSave();
// 				isInitialized = true;
// 				vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
// 			}
// 			catch(err){
// 				// @ts-ignore
// 				vscode.window.showErrorMessage(`${err.message}`);
// 			}
			
// 		});
// 	}
// 	else{
// 		vscode.window.showErrorMessage('intialize: workspacefoleder is underfined.');
// 		return;
// 	}

	
// }



// export function submitProgress(){
// 	if(!isInitialized){
// 		vscode.window.showErrorMessage(`Initialize first to submit`);
// 		return;
// 	}
// 	vscode.window.withProgress({
// 		cancellable: false,
// 		location: vscode.ProgressLocation.Notification,
// 		title: "Submitting Progress..."
// 	},	async (progress) => {
// 		try{
// 			await Git.submitProgress();
// 			vscode.window.showInformationMessage(`Git: Pushed Successfully!`);
// 		}
// 		catch(err){
// 			// @ts-ignore
// 			vscode.window.showErrorMessage(`Git: ${err.message}`);
// 		}
		
// 	});
// }

// export function run(explicit: boolean, Git: GitHandler, Docker: DockerHandler){
// 	// if(!isInitialized){
// 	// 	vscode.window.showErrorMessage(`Initialize first to run`);
// 	// 	return;
// 	// }
	
// 	let notifLoc = explicit? vscode.ProgressLocation.Notification : vscode.ProgressLocation.Window;
	
// 	vscode.window.withProgress({
// 		cancellable: false,
// 		location: notifLoc,
// 		title: "Compiling and Running "
// 	},	async (progress) => {
// 		try{
// 			progress.report({message: "running in docker"});
// 			let response = await Docker.exec('run');
// 			progress.report({message: "Saving Progress"});
// 			await Git.saveProgress();
// 			if(explicit)vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
// 		}
// 		catch(err){
// 			// @ts-ignore
// 			vscode.window.showErrorMessage(err.message);
// 		}
		
// 	});
// // }

// export function onSaveHandler(e: vscode.TextDocument){
// 		console.log('onSaveHandler: save triggered event');
// 		// Giving cooldown of 10 seconds between auto commits on save
// 		// To prevent high amount of commits
// 		if(saveProgresstimerID === null){
// 			run(false);
// 			saveProgresstimerID = setTimeout(()=>{
// 				saveProgresstimerID = null;
// 			}, 10*1000);
// 		}
// }

// export function onFileChangeHandler(){
// 	console.log('onFileChangeHandler: smth changed??\n');
// 	let timerInterval:number = <number> vscode.workspace.getConfiguration().get('Timer Shutdown Interval')
// 	if(timerID === null){
// 		startTime = Date.now();
// 		latestTime = Date.now();
// 		timerID = setTimeout(() => {storeTime(folderPath, latestTime - startTime); timerID = null;}, timerInterval*60*1000);
// 	}
// 	else{
// 		clearTimeout(timerID);
// 		latestTime = Date.now();
// 		timerID = setTimeout(() => {storeTime(folderPath, latestTime - startTime); timerID = null;}, timerInterval*60*1000);
// 		storeTime(folderPath, latestTime - startTime);
// 		startTime = latestTime;
// 	}
// }

// export function intervalForceSave(){
// 	// Save all files in specified time intervals
// 	// This will trigger auto commit on save automatically
// 	let forcedSaveInterval:number = <number> vscode.workspace.getConfiguration().get('Force Save Interval');
// 	forcedSaveTimerID = setTimeout(() => {
// 		vscode.workspace.saveAll(false);
// 		intervalForceSave();
// 	}, forcedSaveInterval*60*1000);
// }

// export function shutDown(){
// 	if(!isInitialized){
// 		vscode.window.showErrorMessage(`Initialize first to shut down`);
// 		return;
// 	}
// 	vscode.window.withProgress({
// 		cancellable: false,
// 		location: vscode.ProgressLocation.Notification,
// 		title: "Shuting Down...."
// 	},	async (progress) => {
// 		try{
// 			let response = await Docker.exec('shut down');
// 			freeSpace();
// 			isInitialized = false;
// 			vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
// 		}
// 		catch(err){
// 			// @ts-ignore
// 			vscode.window.showErrorMessage(`${err.message}`);
// 		}
		
// 	});
// }
// export function freeSpace(){
// 	if(forcedSaveTimerID !== null)clearTimeout(forcedSaveTimerID);
// 	if(fileChangeListenerDisposable !== null)fileChangeListenerDisposable.dispose();
// 	if(fileSaveListenerDisposable !== null)fileSaveListenerDisposable.dispose();
// }

// export function helloWorld(){
// 	if(vscode.workspace.workspaceFolders !== undefined) {
// 		let wf = vscode.workspace.workspaceFolders[0].uri.path ;
// 		let f = vscode.workspace.workspaceFolders[0].uri.fsPath ; 
		

		
// 		let message = `Hello World from AssignmentFetcher! Location is : ${wf} - ${f}` ;
// 		vscode.window.showInformationMessage(message);

// 	} 
// 	else {
// 		let message = "AssignmentFetcher: Working folder not found, open a folder and try again" ;
		
// 		vscode.window.showErrorMessage(message);
// 	}
// } 

src/extension.ts
import * as vscode from 'vscode';
// import { helloWorld } from './actions';
import { sidePanelProvider } from './sidePanelProvider';

// this method is called when your extension is activated
export function activate(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "assignmentfetcher" is now active!');
	
	if(vscode.workspace.workspaceFolders !== undefined){
		const afSidePanel = new sidePanelProvider(context.extensionUri, vscode.workspace.workspaceFolders[0].uri.fsPath );
		
		context.subscriptions.push(vscode.window.registerWebviewViewProvider(sidePanelProvider.viewType, afSidePanel));
		
		let disposable1 = vscode.commands.registerCommand('assignment-fetcher.submit-progress', afSidePanel.submitProgress);
		let disposable2 = vscode.commands.registerCommand('assignment-fetcher.run', afSidePanel.run );
		let disposable3 = vscode.commands.registerCommand('assignment-fetcher.initialize', afSidePanel.initialize);

		
		context.subscriptions.push(disposable1);
		context.subscriptions.push(disposable2);
		context.subscriptions.push(disposable3);
	}

	// let disposable = vscode.commands.registerCommand('assignmentfetcher.helloWorld', helloWorld);
	// context.subscriptions.push(disposable);
}



// this method is called when your extension is deactivated
export function deactivate() {

        // freeSpace();
}



src/gitHandler.ts
import simpleGit, {SimpleGit, SimpleGitOptions} from 'simple-git';
import * as vscode from 'vscode';


export class GitHandler {

    private git: SimpleGit;

    constructor(basePath: string){
        let options: Partial<SimpleGitOptions> = {
            baseDir: basePath,
            binary: 'git',
            maxConcurrentProcesses: 6,
         };
        this.git = simpleGit(options);

    }

    async submitProgress(): Promise<any>;
    async submitProgress(username:string, token: string): Promise<any>;

    async submitProgress(username?:string, token?: string){
        try{
            if(token && username){
                let authRemote: string;
                let repoName: string|undefined = await this.getRepoName();            

                if(repoName !== undefined){
                    authRemote = `https://${username}:${token}@github.com/${username}/${repoName}`;
                    return await this.git.push([authRemote, '-u']);
                }
                else vscode.window.showErrorMessage('Failed to get reponame');
            }
            else{
                return await this.git.push(['origin'])
            }
        }
        catch(err){
            throw(err);
        }
    }


    async saveProgress(){
        try{
            await this.git.add('*').commit('Saving Progress');
            // vscode.window.showInformationMessage('progress saved');
        }
        catch(err){
            throw(err);
        }
    }

    async getRepoName(){
        try{
            let reponame: string|undefined;
            let remotelink = await this.git.getConfig('remote.origin.url');

            if(remotelink.value !== null){
                reponame = remotelink.value.split('/').pop();
                console.log(reponame);
                if(reponame !== undefined && reponame.substr(-4) === '.git')reponame = reponame.substr(0, reponame.length - 4);
                console.log(reponame);
                
                return reponame;
            }
        }
        catch(err){
            throw{err};
        }

    }
}

src/storeHandler.ts
const fs = require("fs");

function jsonReader(filePath: string, cb: any) {
  fs.readFile(filePath, (err: any, fileData: string) => {
    if (err) {
      return cb && cb(err);
    }
    try {
      const object = JSON.parse(fileData);
      return cb && cb(null, object);
    } catch (err) {
      return cb && cb(err);
    }
  });
}

export function storeTime(path: string, activeTimeIncrement: number){
    console.log('storeTime: started');
    activeTimeIncrement /= (1000*60);
    // input was in milliseconds, so converting to minutes
    // Sometimes the values gets updated as '}number' when updating just after assignment
    // is fetched, gets fixed if user clears that mess and rewrites ' "activeTime":0 '  part.
    // why??
    jsonReader(path, (err: any, student: { activeTime: number; }) => {
        if (err) {
          console.log(err);
          return;
        }
        student.activeTime += activeTimeIncrement;
        console.log("Updating");
        fs.writeFile(path, JSON.stringify(student, null, 2), (err: any) => {
          if (err) console.log("Error writing file:", err);
        });
      });
}

module.exports = { storeTime }

src/dockerHandler.ts
import { exec, spawn, ChildProcessWithoutNullStreams } from 'child_process'

const path = require("path");

//  marking stderr to message cause docker compose uses stderr to log the outputs for some reason
//  https://github.com/docker/compose/issues/7346
export type dockerReturn = {
    stdout: string;
    message: string;
    code: number | null;
    signal: string | null;
    time: number;
    timeOut: boolean;
};

export class DockerHandler {

    private workDir: string;
    private containerName: string;

    constructor(workDir: string, containerName: string){
        this.workDir = workDir;
        this.containerName = containerName;
    }

    exec(command: string): Promise<dockerReturn>{
        let process: ChildProcessWithoutNullStreams|null = null;
        const spawnOpts = {
            cwd: this.workDir
        };
        const result: dockerReturn = {
            stdout: 'Docker: ',
            message: 'Docker: ',
            code: null,
            signal: null,
            time: 0,
            timeOut: false,
        };
    	let ymlPath = path.join('.', 'Data', 'docker-compose.yml');
        switch(command){
            case 'start':{
                process = spawn(
                    "docker-compose", 
                    ["-f", ymlPath, "up", "-d"], 
                    spawnOpts
                );
                break;
            }
            case 'run':{
                // running with -it flag gives "the input device is not a TTY" error
                // https://serverfault.com/questions/897847/what-does-the-input-device-is-not-a-tty-exactly-mean-in-docker-run-output
                process = spawn(
                    "docker",
                    ["exec", "-i", this.containerName, "make"],
                );
                break;
            }
            case 'shut down':{
                process = spawn(
                    "docker-compose", 
                    ["-f", ymlPath, "down"], 
                    spawnOpts
                );
                break;
            }
        }

        const begin = Date.now();
        const ret: Promise<dockerReturn> = new Promise((resolve, reject) => {
            
            if(process === null) reject('command not allowed');
            else{
                process.on('exit', (code, signal) => {
                    const end = Date.now();
                    result.code = code;
                    result.signal = signal;
                    result.time = Math.round((end - begin)/100)/10;
                    console.log('Run Result:', result);
                    resolve(result);
                });
        
                process.stdout.on('data', (data) => {
                    result.stdout += data;
                });
                process.stderr.on('data', (data) => (result.message += data));
                process.on('error', (err) => {
                    const end = Date.now();
                    result.code = 1;
                    result.signal = err.name;
                    result.message = `Docker: ${err.message}`;
                    result.time = Math.round((end - begin)/100)/10;
                    console.log('Run Error Result:', result);
                    reject(result);
                });
            }
        });
    
        return ret;


    }

}

src/sidePanelProvider.ts
import * as vscode from 'vscode';
import { storeTime } from './storeHandler';
import { GitHandler } from './gitHandler';
import { DockerHandler } from './dockerHandler';

const path = require("path");
const nconf = require('nconf')

export class sidePanelProvider implements vscode.WebviewViewProvider {
	// we will make an object when the extension activates only
	// so we can store the state variables inside this itself
	// This object wont be destroyed even if the side bar is destroyed
	// since its not recreated, the html rendering will only be done once(at activation)
	// so injecting contents to getHtml wont work, can do via script only

	public static readonly viewType = 'assignmentfetcher.panelView';
	private _view?: vscode.WebviewView;

	private isLoading: boolean; 
	private isInitialized: boolean; 
	private Git: GitHandler;
	private Docker: DockerHandler;
	private folderPath: string;
	private startTime:number;
	private latestTime:number;
	private timerID:null|NodeJS.Timeout = null;
	private saveProgresstimerID:null|NodeJS.Timeout = null;
	private forcedSaveTimerID:null|NodeJS.Timeout = null;

	private fileSaveListenerDisposable: vscode.Disposable|null = null;
	
	constructor(private readonly _extensionUri: vscode.Uri, folderPath: string){ 
		this.isLoading = false;
		this.isInitialized = false;
		this.startTime = Date.now();
		this.latestTime = Date.now();
		this.folderPath = folderPath;
		this.Git = new GitHandler(this.folderPath);
		nconf.use('file', { file: path.join(this.folderPath, 'Data', 'metadata.json') });
		nconf.load();
		this.Docker = new DockerHandler(this.folderPath, nconf.get("container_name"));
	}

	public resolveWebviewView(
		webviewView: vscode.WebviewView,
		context: vscode.WebviewViewResolveContext,
		_token: vscode.CancellationToken,
	) {
		this._view = webviewView;
		webviewView.webview.options = {
			// Allow scripts in the webview

			enableScripts: true,
			localResourceRoots: [
				this._extensionUri
			]
		};

		webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);
		webviewView.webview.onDidReceiveMessage(data => {
			console.log(`frontend to backend ${data.type}`);
			switch (data.type) {
				case 'initialize':{
					this.initialize();
					break;
				}
                case 'submitProgress':{
					this.submitProgress();
					break;
                }
				case 'run':{
					this.run(true);
					break;
                }

				case 'getState':{
                    this.sendState();
					break;
                }
				default: {
                    console.error('Unknown event received from webview');
                }
			}
		});
	}

	private sendState = ()=>{
		this._view?.webview.postMessage({type:'getState', isInitialized: this.isInitialized});
	}

	public initialize = ()=>{
		console.log( "state", this.isInitialized, this.isLoading);
		console.log('backend: initialize');
		if(this.isLoading) return;
		this.isLoading = true;
		// this.isInitialized = !this.isInitialized;
		if(!this.isInitialized){
				vscode.window.withProgress({
					cancellable: false,
					location: vscode.ProgressLocation.Notification,
					title: "Starting Container"
				},	async (progress) => {
					try{
						let response = await this.Docker.exec('start');
						this.fileSaveListenerDisposable = vscode.workspace.onDidSaveTextDocument(this.onSaveHandler);
						
						this.intervalForceSave();
						this.isInitialized = true;
						this.sendState();
						vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
						this.isLoading = false;
					}
					catch(err){
						// @ts-ignore
						vscode.window.showErrorMessage(`${err.message}`);
						this.isLoading = false;
					}
					
				});
		}
		else{
			vscode.window.withProgress({
				cancellable: false,
				location: vscode.ProgressLocation.Notification,
				title: "Shuting Down...."
			},	async (progress) => {
				try{
					let response = await this.Docker.exec('shut down');
					this.freeSpace();
					this.isInitialized = false;
					this.sendState();
					vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
					this.isLoading = false;
				}
				catch(err){
					// @ts-ignore
					vscode.window.showErrorMessage(`${err.message}`);
					this.isLoading = false;
				}
			});
		}
	}

	public submitProgress = ()=>{
		console.log('backend: submit');
		// if(!this.isInitialized){
		// 	vscode.window.showErrorMessage(`Initialize first to submit`);
		// 	return;
		// }
		vscode.window.withProgress({
			cancellable: false,
			location: vscode.ProgressLocation.Notification,
			title: "Submitting Progress..."
		},	async (progress) => {
			try{
				await this.Git.submitProgress();
				vscode.window.showInformationMessage(`Git: Pushed Successfully!`);
			}
			catch(err){
				// @ts-ignore
				vscode.window.showErrorMessage(`Git: ${err.message}`);
			}
			
		});
	}

	

	public run = (explicit: boolean = true)=>{
		console.log('backend: run');

		// if(!isInitialized){
		// 	vscode.window.showErrorMessage(`Initialize first to run`);
		// 	return;
		// }
		let notifLoc = explicit? vscode.ProgressLocation.Notification : vscode.ProgressLocation.Window;
		vscode.window.withProgress({
			cancellable: false,
			location: notifLoc,
			title: "Compiling and Running "
		},	async (progress) => {
			try{
				progress.report({message: "running in docker"});
				let response = await this.Docker.exec('run');
				progress.report({message: "Saving Progress"});
				await this.Git.saveProgress();
				if(explicit)vscode.window.showInformationMessage(`${response.message} (${response.time} sec)`);
			}
			catch(err){
				// @ts-ignore
				vscode.window.showErrorMessage(err.message);
			}
			
		});
	}


    private _getHtmlForWebview(webview: vscode.Webview) {
		// Get the local path to main script run in the webview, then convert it to a uri we can use in the webview.
		const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.js'));

		// Do the same for the stylesheet.
		const styleResetUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'reset.css'));
		const styleVSCodeUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'vscode.css'));
		const styleMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'media', 'main.css'));

		// Use a nonce to only allow a specific script to be run.
		const nonce = getNonce();

		return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<!--
					Use a content security policy to only allow loading images from https or from our extension directory,
					and only allow scripts that have a specific nonce.
				-->
				<meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src ${webview.cspSource}; script-src 'nonce-${nonce}';">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${styleResetUri}" rel="stylesheet">
				<link href="${styleVSCodeUri}" rel="stylesheet">
				<link href="${styleMainUri}" rel="stylesheet">
				
				<title>Assignment Fetcher</title>
			</head>
			<body>
				<label class="switch">
					<input id="check" type="checkbox">
					<div></div>
					<span id= "indicator">OFF</span>
				</label>
				<div id ="actions">
					<button id="submitProgress">Submit</button>
					<button id="run">Run</button>
				</div>
				<script nonce="${nonce}" src="${scriptUri}"></script>
			</body>
			</html>`;
	}
	private freeSpace = ()=>{
		if(this.forcedSaveTimerID !== null)clearTimeout(this.forcedSaveTimerID);
		// if(this.fileChangeListenerDisposable !== null)this.fileChangeListenerDisposable.dispose();
		if(this.fileSaveListenerDisposable !== null)this.fileSaveListenerDisposable.dispose();
	}
	
	private intervalForceSave = ()=>{
		// Save all files in specified time intervals
		// This will trigger auto commit on save automatically
		let forcedSaveInterval:number = <number> vscode.workspace.getConfiguration().get('Force Save Interval');
		this.forcedSaveTimerID = setTimeout(() => {
			vscode.workspace.saveAll(false);
			this.intervalForceSave();
		}, forcedSaveInterval*60*1000);
	}
	
	//  this listener didnt get called wehn I used in special character heavy path, so combining this with onSave listener for now
	private onFileChangeHandler = ()=>{
		console.log('onFileChangeHandler: smth changed??\n');
		let timerInterval:number = <number> vscode.workspace.getConfiguration().get('Timer Shutdown Interval')
		let storePath: string = path.join(this.folderPath, 'Data', 'studentData.json');

		if(this.timerID === null){
			this.startTime = Date.now();
			this.latestTime = Date.now();
			this.timerID = setTimeout(() => {storeTime(storePath, this.latestTime - this.startTime); this.timerID = null;}, timerInterval*60*1000);
		}
		else{
			clearTimeout(this.timerID);
			this.latestTime = Date.now();
			this.timerID = setTimeout(() => {storeTime(storePath, this.latestTime - this.startTime); this.timerID = null;}, timerInterval*60*1000);
			storeTime(storePath, this.latestTime - this.startTime);
			this.startTime = this.latestTime;
		}
	}
	
	// have to assign function like, name = () =>{}.. else 'this' keyword will not refer
	// to our object but smth else and it will crash
	private onSaveHandler = (e: vscode.TextDocument)=>{
		console.log('onSaveHandler: save triggered event '); 
		// Giving cooldown of 10 seconds between auto commits on save
		// To prevent high amount of commits
		if(this.saveProgresstimerID === null){
			this.run(false);
			this.saveProgresstimerID = setTimeout(()=>{
				this.saveProgresstimerID = null;
			}, 10*1000);
		}
		this.onFileChangeHandler();
	}
}

function getNonce() {
	let text = '';
	const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
	for (let i = 0; i < 32; i++) {
		text += possible.charAt(Math.floor(Math.random() * possible.length));
	}
	return text;
}

